Implement Snackbar notifications for new releases

1zaman opened the pull request:
### Description

[MA-1995][]

An OkHttp interceptor has been added to capture the headers notifying of new versions of the app that have been released on the app stores. These are broadcasted over the event bus and displayed on `MyCoursesListActivity` (the main dashboard Activity) in the form of Snackbars, that will persist until the user swipes them away.

The Activity is treated as a session, with notifications not being repeated in a single session, but reset after the session is closed. This design is so that the notifications are always provided directly by the server, instead of being persisted on the local device. This will prevent wrongly sent headers to permanently change the app state to an incorrect one.

There are three different notifications in descending order of priority. New notifications will only be shown if they have a higher priority than any other one that has been shown in the current session.

1. Unsupported app version notifications. This is when we get a 426 (upgrade required) error as a response. This means that the API we're calling has been removed, and will no longer deliver anything.

2. Deprecation notification. This is when we get notified of a new version of the app, along with a deadline after which the current version won't have it's API calls supported any longer. These are prioritized according to the closeness of their deadlines.

3. New release notification. This is when we get notified of a new version of the app, without any deadline for support of the current version.  These are prioritized according to the largest new version number.

For comparing versions, a custom `Version` class has been written to strictly represent the app scheme, which can be defined as 'X.X.X.Y', where X represents numbers, and Y represents an optional alphanumeric string. Y is ignored for comparison purposes, since it should always coincide with a change in the X portion. The Version class simply splits the string by the period token, takes the first three (or less) substrings it finds, and converts them into numbers. If any of them
can't be converted into an unsigned integer, then it throws a `ParseException` (upon encountering which the `NewVersionBroadcastReceiver` implementation terminates it's operations). A test suite for the `Version` class has been written as well.

The Snackbar has an action button for updating the app, which attempts to open the app's section in an app store, using a list of URIs that are defined in the config. If the initial app store that was used to install the app can be resolved through one of the URIs, then it's given priority, and attempted to launch directly. Otherwise, all the URIs are attempted in order until one of them resolves to a registered Activity. If none are found, then the action button is not shown. If the app store was somehow uninstalled after displaying the update button, then it results in a toast error message.

Unit tests have been added for `NewVersionAvailableEvent`, and the `NewVersionBroadcastInterceptor`, adding functional, validity, and comparison assertions for all possible permutations.

The event bus posting verification tests (for `NewVersionBroadcastInterceptor` and the static post() method) unfortunately doesn't work in the parameterized test suite at the moment, since the event bus initialization requires loading the Android framework, and we currently don't have a parameterized Robolectric test runner implementation. Nor can RoboGuice be used to inject a mock instance of the event bus, since initializing RoboGuice also requires a Context (which is only available in the Android framework). Therefore, these tests are being ignored for now.

No tests have been written for verifying the event bus registration at the UI level, and the subsequent display of the appropriate Snackbar notification, since there is currently no official shadow implementation for Snackbar in the Robolectric framework. If a custom Snackbar shadow is added to the project at some point, then this test can be added as well at that point.

### Supplementary Changes:

* [Add a list of constants for HTTP status codes](https://github.com/edx/edx-app-android/pull/748/commits/383dc9c360652bc562fc12c85720e342a14cd191)
* [Support parsing dates according to ISO 8601](https://github.com/edx/edx-app-android/pull/748/commits/41b6a71e0bc0f2a6d6b3a4baeb8eaa2c1d4ec15f)
* [Update AssertJ Core from 1.x to 2.x](https://github.com/edx/edx-app-android/pull/748/commits/0a7e7121b1d16334ad058d2cec5c9e9a210a1f21)

### Testing
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @bguertin
- [x] Code review: @mdinino
- [x] Code review: @miankhalid
- [ ] Code review: @BenjiLee
- [x] UI strings/error msgs review: @catong
- [x] UX review: @lizcohen

[MA-1995]: https://openedx.atlassian.net/browse/MA-1995

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/MyCoursesListActivity.java:
```diff
@@ -79,4 +118,34 @@ public void onFail(Exception ex) {
+             * text also takes up lots of space at the right. Manually increase
+             * the maximum lines limit to 10 in order to have our current messages
+             * displayed.
+             * TODO: Make the strings shorter?
```
@catong: Any suggestions on shorter strings for the Snackbar messages?

catong replied:
@1zaman I will take a look and tag you on the Version Upgrade Alerts wiki. Thanks!

lizcohen commented:
<img width="468" alt="screen shot 2016-06-15 at 10 33 25 am" src="https://cloud.githubusercontent.com/assets/19694600/16084148/ad3ee3ee-32e5-11e6-9e62-909d89cb51e5.png">
<img width="418" alt="screen shot 2016-06-15 at 10 40 36 am" src="https://cloud.githubusercontent.com/assets/19694600/16084149/ad4558dc-32e5-11e6-8275-e5bc4af36386.png">

Make sure to watch the snackbar design best practices: https://material.google.com/components/snackbars-toasts.html#snackbars-toasts-specs

We should really try to get all of the text in snackbars down to 2 lines.

1zaman commented:
@lizcohen: We're using the Design Support Library for showing the Snackbars, which presumably already follows the Material design specs. I agree that we should try to shorten the message strings to try to get it to fit into two lines.

Note that we can never guarantee that the text won't be truncated on any maximum line limit, as screen and font sizes are variable, and the string will also be localized into other languages in the future.

catong commented:
@1zaman @lizcohen 
For these message given that there's an Upgrade link right there, we can drop the "Upgrade to the Latest Version" part of the text.
For Unsupported: "Your version of the app is no longer supported."
For Outdated: "A new version of the app is available."

I'm still wrestling with Deprecated. @1zaman can you help me understand why the message needs to be different from "Unsupported"? If the app is already no longer supported, what happens on the deadline date? I'm not clear why we need to warn about being able to access course content (wouldn't the Unsupported case also have the same limitation?)

catong commented:
Ah, I see a description of Deprecated: "This is the version that would contain critical/major application changes and would be bound with a deadline."
How about something like: "Your version of the app is not supported after (date)."
Or, "To continue accessing course content, upgrade to the latest version of the app by (date)."
@1zaman @lizcohen Your thoughts?

lizcohen commented:
@catong @1zaman 
I don't think we should ever include a specific date. If it is already deprecated, then the date will have passed. Can we say something like "Your version of the app is no longer supported and will have limited functionality."?

catong commented:
@lizcohen @1zaman 
Would a message that is identical to the Unsupported case work for Deprecated? "Your version of the app is no longer supported."

verdanmahmood commented:
@lizcohen @catong
I think we should include the date in the notification in case of Deprecated Version, because we will notify user that a new version of app is available but he can continue using existing version to a certain deadline (that is configurable) and after that time period he is required to either upgrade app or continue using in offline mode.

catong commented:
Thanks @verdanmahmood Is the difference between the Deprecated and Unsupported cases that for Unsupported there is no deadline and users can continue to access the course indefinitely with the older app version?

verdanmahmood commented:
@catong here is the doc with the definition of three types of notifications related to versions. 
https://openedx.atlassian.net/wiki/display/MA/Version-based+App+Upgrade

verdanmahmood commented:
@catong yes, but with cached data only. We'll not fetch the new data from server. and yes there won't be any date in case of Unsupported version. User would be able to access cached data indefinitely.

catong commented:
@verdanmahmood So unless I'm missing something, the behavior is the same for both Deprecated and Unsupported: - user can access cached data indefinitely if they don't upgrade, regardless of whether there is a deadline.

verdanmahmood commented:
@catong nope, in case of deprecated version (which is still deprecating, not yet fully deprecated), we are encouraging user that new version is available, you should upgrade to new version, and current version will be Unsupported after a certain date. App will run normally in this case, we will still be supporting this app, with just a notification for a reminder. 

Unsupported version is when the deadline from deprecated version will be passed, and the app will not be able to run properly. This is the time we'll only show cached data to user and for new data, we'll ask him to upgrade to latest version, because we are not supporting this version of the app.

1zaman commented:
@catong: I like your suggested string of "Your version of the app is not supported after (date).", although maybe "is not" should be "won't be". It's similar to the unsupported version message, and we can try to ensure that the date is kept as small as possible (i.e. maybe remove the year if it's the same one).

catong commented:
So maybe this for Unsupported and Deprecated, though it might be longer than 2 lines: "Your version of the app is no longer supported. You can access only cached content".
@lizcohen @1zaman ?

lizcohen commented:
Do we need the line "You can access only cached content"? I think we can leave it at "Your version of the app is no longer supported."

catong commented on VideoLocker/res/values/errors.xml:
```diff
@@ -63,6 +63,19 @@
     <string name="video_offline_message">Please select a video that you’ve downloaded onto your device.</string>
 
+    <!-- New version available -->
+    <!-- Snackbar notification shown when the current version of the app is unsupported -->
```
@1zaman Please update strings for app_version_unsupported, app_version_deprecated, and app_version_outdated per the conversation [here](https://github.com/edx/edx-app-android/pull/748#issuecomment-226272349).

1zaman replied:
✔️

1zaman commented:
Deprecated (after deadline) = Unsupported

The [wiki][Version-based App Upgrade] says we should be displaying the last supported date in the deprecation notification:

> For Case 2:
> API response contains a header 'EDX-APP-LATEST-VERSION' and 'EDX-APP-VERSION-LAST-SUPPORTED-DATE'; notify user that a latest version of app is available and he is required to upgrade app before  'EDX-APP-VERSION-LAST-SUPPORTED-DATE'.

If we're not using the date anywhere, then there doesn't seem to be much point in the `EDX-APP-VERSION-LAST-SUPPORTED-DATE` header. Anyway, this should probably be discussed on the wiki, and if the consensus is to not show the date, then the wiki guidelines should be updated.

[Version-based App Upgrade]: https://openedx.atlassian.net/wiki/display/MA/Version-based+App+Upgrade

1zaman commented on VideoLocker/res/values/errors.xml:
```diff
@@ -63,6 +63,19 @@
+    <string name="app_version_upgrade_button">Upgrade</string>
+    <!-- Toast notification for when the user presses the upgrade action button from the Snackbar
+         notifications while no app store or browser is installed on the device -->
+    <string name="app_version_upgrade_app_store_unavailable">Play Store not found on device.</string>
```
@catong: Any suggestions for updating this last string, or are you fine with it as it is?

catong replied:
@1zaman We should add Google: "Google Play Store not found on device."

1zaman replied:
@catong: Note that I have now changed this to "No supported app store found on device.", since the actual app store we're directing to has been [made][1] configurable (and would not necessarily always be the Play Store). Can you confirm this phrasing, or do you want to suggest some edits to it?

[1]: https://github.com/edx/edx-app-android/pull/748#discussion_r73464816

catong replied:
@1zaman That sounds perfect. Thanks!

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/AppStoreUtils.java:
```diff
@@ -0,0 +1,60 @@
+            context.startActivity(intent);
+        } catch (ActivityNotFoundException e) {
+            intent.setData(Uri.parse("https://play.google.com/store/apps/details?id=" +
+                            BuildConfig.APPLICATION_ID));
```
@BenjiLee, @bguertin: Do we want to have a config for third party customizations to specify their own app store URI, in case their preferred distribution channel is not the Play Store? The `market://` URI can be presumably handled by any app store, but we might prefer to directly try a specific one first. Similarly, if no app store is available, then the web page that we open could be customized to be something other than the Play Store.

BenjiLee replied:
I'm going to go with yes in the spirit of being open source. I know some countries can't use the play store and some countries have their own system for upgrading the app.

@scottrish  FYI

bguertin replied:
Yes, we should definitely have a config flag for the app store URL.

Additionally, for the official edX app, we should explicitly open the Play Store app. Users can install other apps that intercept "market://" URLs, but in our case we don't want them to open a different app store because we know they need to use Google Play. See: http://stackoverflow.com/a/28090925 for a potential solution.

1zaman replied:
Hmm, so in that case, maybe we should have a list of supported app stores in the config, which we can try to invoke sequentially, and tell the user to install the first one if none of them is available (and there's no web browser installed either). We might also start distributing the app in other app stores in the future, and should be able to open any supported store that the user has already installed on the device.

Note that the Play Store could be specifically opened by just trying to open the HTTP URL of the app in the Play Store website, since it captures those too if it's installed (and other stores shouldn't be capturing it's specific URLs). This also has the advantage of being a generic technique that should work for any app store without having to also check their package names.

bguertin replied:
Let's not solve for apps intercepting the `market://` URI, but lets definitely make both of these configurable. How about just a config for the two URIs (name them whatever you want). E.g.:
```
APP_STORE_NATIVE_URI: 'market://details?id={applicationId}'
APP_STORE_WEB_URI: 'https://play.google.com/store/apps/details?id={applicationId}'
```
We'll try the native URI first, and fill in the `{applicationId}` placeholders, if present.

bguertin replied:
And perhaps, also have a method `boolean canOpenAppInAppStore()` that returns true if either of the URIs are resolvable. And then only show the "Update" link if it returns true. Make sense?

1zaman replied:
> Let's not solve for apps intercepting the `market://` URI, but lets definitely make both of these configurable.

Done in commit 3e85fbcf6f164567e8885a433677caca9bd72503. The name and structure of the flags still needs to be finalized in the wiki (and then added in the existing configs), but I have tentatively assumed a `APP_UPDATE_URIS` top-level object containing `NATIVE` and `WEB` fields.

> And perhaps, also have a method `boolean canOpenAppInAppStore()` that returns true if either of the URIs are configured and resolvable. And then only show the "Update" link if it returns true. Make sense?

Done in commit 3864c8c5da4bdf984de9b495d2559179ca33d21b.

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,189 @@
+     */
+    private NewVersionAvailableEvent(@Nullable final ArtifactVersion newVersion,
+                                    @Nullable final Date lastSupportedDate,
+                                    final boolean isUnsupported) {
```
Autoformat 😉

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/http/NewVersionBroadcastInterceptor.java:
```diff
@@ -0,0 +1,75 @@
+             * has improved support for ISO 8601, including UTC offsets and the
+             * "Z" notation.
+             * TODO: Implement this all over the app instead of the existing
+             * SimpleDateFormat.
```
SimpleDateFormat should appear in the first line as well, for Android Studio to read and display it properly.

1zaman replied:
That's an issue with the Android Studio highlighter, in that it only highlights the line that include the "TODO" keyword. I don't think we need to restrict all our Todos to one line because if this 😉

miankhalid replied:
If we can accommodate our TODO to a single line, why not do it 😉

1zaman replied:
Cause I don't care about it 😉

bguertin replied:
What format is the server giving for the date? I can't imagine any reason why SimpleDateFormat would not suffice, we use it for other date strings in the app.

1zaman replied:
It's sending a `+00:00` UTC offset instead of the `Z` character, which means zero offset. `SimpleDateFormat` can parse these via the `XXX` pattern, but it doesn't support one pattern to parse both of these. The `FastDateFormat` in Apache Commons Lang 3 offers the `ZZ` pattern that does parse both of them, amongst other benefits such as thread-safety and caching. It doesn't provide full support for parsing ISO 8601 compliant date formats though, since that requires an actual parser instead of just a formatter, in order to support optional parts (e.g. fractional seconds).

I guess it would be better to use an actual parsing library for ISO 8601. I didn't find any during my initial web search on this, but searching again now I found [ISO 8601 DateFormat][], which provides a parser for ISO 8601 that also implements the `DateFormat` interface. It also links to [ThreeTen Android Backport][], which backports the java.time.* [package][java.time.* package] from Java 8 (which is a full replacement of the old `Date` and `Calendar` APIs). I suppose we should be using one of these alternatives for parsing ISO 8601 dates. What do you think?

[ISO 8601 DateFormat]: https://github.com/nobuoka/java-date-format-iso8601
[ThreeTen Android Backport]: https://github.com/JakeWharton/ThreeTenABP
[java.time.* package]: https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html

brianguertin replied:
> doesn't support one pattern to parse both of these

Why do we need to only use one pattern? We can easily use a list of patterns that we expect from the server and try them all. This won't be a bottle-neck in our app by any means.

1zaman replied:
> in fact we already have a `DateUtil` with a `convertToDate(String)` method, though it handles failure silently which is not great

It's not being used in the `ServerJsonDateAdapter` though, which is our custom Gson type adapter for `Date` type. In fact, it only seems to be used in the course details model, which parses to `String` instead of `Date` 😦

1zaman replied:
So, it looks like the `XXX` pattern was only introduced in Java 7, which is only supported in KitKat+. Fortunately, however, I found out that Gson 2.5 imported an ISO 8601 parser from Jackson, so I upgraded Gson to 2.7 and changed our `DateUtils` class to call through to it instead in commit 2e7fb35587e810b98cd46bb75936f957384667d3. I also changed our date type adapter to use that, since Gson still doesn't include a dedicated type adapter for ISO 8601 dates in the main library (it only falls back to it on deserialization, but formats to a specified or default date/time pattern using the device's local time zone when serializing 😞).

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/view/MyCoursesListActivity.java:
```diff
@@ -32,10 +42,39 @@ protected void onCreate(Bundle savedInstanceState) {
+         * probably set up the CoordinatorLayout as a global content view parent in
+         * BaseFragmentActivity.
+         */
+        coordinatorLayout = new CoordinatorLayout(this);
```
Maybe add a TODO here to remove this logic once we switch to using `CoordinatorLayout` all over the app.

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/view/MyCoursesListActivity.java:
```diff
@@ -2,21 +2,31 @@
+
 public class MyCoursesListActivity extends BaseSingleFragmentActivity {
 
+    private CoordinatorLayout coordinatorLayout;
```
`@NonNull` can be added here.

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/view/MyCoursesListActivity.java:
```diff
@@ -79,4 +118,34 @@ public void onFail(Exception ex) {
+             * TODO: Make the strings shorter?
+             */
+            ((TextView) snackbar.getView().findViewById(
+                    android.support.design.R.id.snackbar_text)).setMaxLines(7);
```
Discrepancy between the set maxLines in comment (10) and here (7) 😉

1zaman replied:
Yeah, 7 lines are sufficient to display all the error messages on my device, but I thought I would set it to a nice round 10, but forgot about it. It looks like we're trying to make the strings short enough to fit in 2 lines anyway, so maybe we won't need this now.

miankhalid commented:
Done with the first pass. Still need to test the various scenarios on an actual device.

bguertin commented on VideoLocker/build.gradle:
```diff
@@ -60,6 +60,7 @@ dependencies {
     // with module updates.
     compile project (':android-iconify-fontawesome')
     compile fileTree(dir: 'libs', include: '*.jar')
+    compile 'org.apache.commons:commons-lang3:3.4'
```
This is a large dependency to drag in. Why exactly do we need it?

1zaman replied:
We were already importing version 2 of the Apache Commons Lang library as a transitive dependency of the Commons IO library, and it has lots of convenient utilities. and the `FastDateFormat` in version 3 had increased support for parsing ISO 8601 UTC offsets, so I thought that it would be a net benefit to import and use the latest version of it. Utilities that aren't utilized in the code base would be stripped out by ProGuard anyway. However, per my previous [comment][], we should probably use a dedicated parser for ISO 8601 instead, so I suppose we can leave out this dependency for now.

[comment]: https://github.com/edx/edx-app-android/pull/748#discussion_r67686918

1zaman replied:
OK, so I have [removed][1] the `FastDateFormat` dependency in favor of using Gson 2.5's implementation of an ISO 8601 parser. I have also removed all usages of the Apache Commons Lang 2.4 dependency, which was actually a compile-time only transitive dependency of the RoboGuice annotation preprocessor RoboBlender, and caused a `NoClassDefFoundError` upon runtime usage, as [discussed][2] in #750. The utility classes that we were using have been imported in the project itself.

[1]: https://github.com/edx/edx-app-android/pull/748#discussion_r69355348
[2]: https://github.com/edx/edx-app-android/pull/750#issuecomment-227444716

bguertin replied:
So, can this dependency line be removed now?

1zaman replied:
Yeah, this import was accidentally readded by commit efb2bf4b3469d4eab136be2c94df36c56ad1d70d. Removed in commit 2fc30396ca0a332631e0eaf779e0f1074b907fa1.

catong commented on VideoLocker/res/values/errors.xml:
```diff
@@ -63,6 +63,19 @@
+    <!-- Snackbar notification shown when the current version of the app is outdated -->
+    <string name="app_version_outdated">A new version of the app is available. Upgrade to the latest version.</string>
+    <!-- Snackbar action button label for opening an app store to upgrade the app to the latest version -->
+    <string name="app_version_upgrade_button">Upgrade</string>
```
@brianguertin pointed out that the more accurate term is "Update" to a version, rather than "Upgrade". If this button is used only for these types of version changes, can we change the button name here to "Update"? @1zaman @verdanmahmood @lizcohen FYI @mikekatz

1zaman replied:
> Not sure if there are additional places where "upgrade" should be replaced by "update".

No, these are the only strings that are being used in the Snackbar messages.

catong replied:
Thanks @1zaman, so if these buttons are applied only to the Snackbar messages, please change the button to read "Update" instead of "Upgrade".

1zaman replied:
Already been done in commit eb6b4d8eb6ea93c9f5a4ea6d62f2f85cad589abb.

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,189 @@
+     *                          each other due to wrong local clock time or an inconsistency in the
+     *                          server configurations).
+     */
+    public static void post(@Nullable final ArtifactVersion newVersion,
```
It's not immediately obvious to me (even after reading the doc) what happens if you pass in null to these variables? How is the behavior impacted?

1zaman replied:
There are three properties here, each corresponding to one of the notification types that I described in the summary, and each also corresponding to a header, or an error code in the case of `isUnsupported`. Any of these may be `null` (or `false`), in which case it's automatically marked as a type corresponding to the property with the highest priority that is available. I suppose it would not be a valid event if all three properties are `null` or `false`, and I'll add a check for that in the constructor and here.

1zaman replied:
I have the constructor to validate that there is _some_ data to show, and added the details in the javadocs as well in commit b946fa1352dbfb5bd6cea4030d6f1b8778885782. The helper `post()` method simply returns without doing anything if it catches an `IllegalArgumentException` upon instantiating the event.

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,189 @@
+     *                          each other due to wrong local clock time or an inconsistency in the
+     *                          server configurations).
+     */
+    private NewVersionAvailableEvent(@Nullable final ArtifactVersion newVersion,
```
same comment here

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * @return The version number of the latest release of the app.
+     */
+    @Nullable
```
Similar comments here. What does a null return denote?

mdinino commented:
I had small comments. Other than those 👍

BenjiLee commented:
@1zaman I'll get to this first thing tomorrow USA time.

BenjiLee commented:
Is temp-lahore sandbox down?

BenjiLee commented:
Done with first pass

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/util/AppStoreUtils.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public final class AppStoreUtils {
+    // Make this class non-instantiable
+    private AppStoreUtils() {
```
pro-tip: use an `enum` to shorten this into a single semi-colon!
e.g.
```java
public enum AppStoreUtils {
   ;
   // methods
}
```

1zaman replied:
It adds extra static methods for enums though (`values()`, `valueOf()`) 😞

bguertin replied:
Fair point, but you're add one extra method by doing it this way. I guess I prefer the shorter code.

Also, using `enum` additionally prevents `newInstance()` from working, where-as it will still work if you just use a private constructor.

bguertin replied:
/pointless argument :laughing:

1zaman replied:
> Fair point, but you're add one extra method by doing it this way. I guess I prefer the shorter code.

Yeah, the constructor needs to be explicitly defined as private, but making it a standard class removes those pesky extra methods in the public API.

> Also, using `enum` additionally prevents `newInstance()` from working, where-as it will still work if you just use a private constructor.

Which is why I implemented it to throw an `UnsupportedOperationException` upon instantiation 😉

> /pointless argument :laughing:

Agreed.

1zaman commented:
I have addressed all the comments, except for [making][1] the preferred app store configurable, which might merit further discussion, and would require adding new fields in the config.

@catong, @lizcohen: The new Snackbar notification strings look like this:

Unsupported              | Deprecated/Outdated
-------------------------|----------------------
![version unsupported][] | ![version outdated][]

[1]: https://github.com/edx/edx-app-android/pull/748#discussion_r67304967

[version unsupported]: https://cloud.githubusercontent.com/assets/4244109/16534247/c4958ede-3ff7-11e6-882e-56c75e52ba87.png
[version outdated]: https://cloud.githubusercontent.com/assets/4244109/16534256/d7968dbc-3ff7-11e6-811a-df0d50d17492.png

1zaman commented:
Note that there seems to be some issue with the sandbox not being immediately updated upon changing the app version config. It might be due to caching.

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,193 @@
+     */
+    @Override
+    public int compareTo(@NonNull final NewVersionAvailableEvent another) {
+        int result = BooleanUtils.compare(isUnsupported, another.isUnsupported);
```
Can be replaced with `Boolean.valueOf(isUnsupported).compareTo(another.isUnsupported)`

1zaman replied:
I am now already performing some explicit comparison logic for `null` objects in commit 4a4bd30be8b792670912d812c9b466b1aa7781bf as per your [comment][], so we might as well do the boolean comparison logic manually as well. It actually simplifies the logic, and since we're already returning the values explicitly now there's no harm in doing it here as well.

[comment]: https://github.com/edx/edx-app-android/pull/748#discussion_r69955691

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,193 @@
+     * @return a negative integer if this instance has lesser priority than {@code another};
+     *         a positive integer if this instance has greater priority than {@code another};
+     *         0 if this instance has the same priority as {@code another}.
+     * @throws ClassCastException if {@code another} cannot be converted into
```
this method does not throw `ClassCastException`

1zaman replied:
Good point. I used the javadoc from the `Comparable<T>` as a template which doesn't have it's generic type defined, so it has this clause. As long as you're calling this specific method as defined without unsafely casting it to a non-generic `Comparable` type, it should only ever receive the instances of the same class as the parameter in the `compareTo()` method, and thus no `ClassCastException` should ever be thrown. I have removed this javadoc declaration in commit ee0d62e006addbb4334b5490edc5da0cf3a58282.

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,193 @@
+             */
+            result = 0 - ObjectUtils.compare(lastSupportedDate, another.lastSupportedDate, true);
+            if (result == 0) {
+                result = ObjectUtils.compare(newVersion, another.newVersion);
```
It's not worth importing these classes to avoid a couple null checks here.

Actually, I think this would be more legible if we had the null inline anyway.

1zaman replied:
> I think this would be more legible if we had the null inline anyway.

Done in commit 4a4bd30be8b792670912d812c9b466b1aa7781bf.

> It's not worth importing these classes to avoid a couple null checks here.

Imports reverted in commit efb2bf4b3469d4eab136be2c94df36c56ad1d70d, and new ticket created for fixing the existing references to the compile-only Commons Lang 2.4 dependency at [MA-2682][].

[MA-2682]: https://openedx.atlassian.net/browse/MA-2682

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/third_party/versioning/ComparableVersion.java:
```diff
@@ -0,0 +1,474 @@
+ * @author <a href="mailto:kenney@apache.org">Kenney Westerhof</a>
+ * @author <a href="mailto:hboutemy@apache.org">Hervé Boutemy</a>
+ */
+public class ComparableVersion
```
Totally optional suggestion: Do we need to support more than just comparing the numbers? I have a simple class here we could use, it only compares numbers: https://gist.github.com/brianguertin/ada4b65c6d1c4f6d3eee3c12b6ce021b

1zaman replied:
Since the version name can technically be any string, and our existing format is merely a convention, it seems better to be able to perform some kind of reasonable comparison for any possible format. Your code would crash with a `NumberFormatException` if it ever got a non-number character in the version string. If we were getting the version code as well then it would have been a simple numeric comparison, but unfortunately the version code is probably specific to Android and is not being delivered by the server.

bguertin replied:
> Since the version name can technically be any string.

Is that true, I thought we would be using [semantic versioning](http://semver.org/)

> Your code would crash with a NumberFormatException if it ever got a non-number character in the version string.

That is by design; if there is a non-numeric character, it is not a valid semantic version string.

bguertin replied:
So, the version format is described precisely here: https://github.com/edx/edx-platform/blob/5f572dfc674ae576ee72f4308d18b470e7498891/lms/djangoapps/mobile_api/models.py#L42

It says: `Version should be in the format X.X.X.Y where X is a number and Y is alphanumeric`

I added a test for the alphanumberic case (which differes from SemVer) to this Version class: https://gist.github.com/bguertin/40d1b00e11adeaa1357e4ab747882d30

1zaman replied:
If we can get an alphanumeric character, then that should be part of the sorting as well (in order to get deterministic comparison if we get two versions that differ only by that last character), and the `ComparableVersion` implementation seems to be providing reasonable heuristics for this comparison.

bguertin replied:
> in order to get deterministic comparison if we get two versions that differ only by that last character

It's equally deterministic to consider them to be equal.

> reasonable heuristics

Our API does not define semantics for ordering the alphanumeric portion, and in fact ignores the alphanumeric portion in it's own comparison logic. Therefore, it is correct for us to ignore it in the client as well.

1zaman replied:
OK, I have added your implementation with some tweaks and fixes and added documentation in commit 10eddf9e93d806857812c66ed5e4957ab326ef2d.

Note that the specs for Semantic Versioning also allow adding alphanumeric pre-release version identifiers, with a specific (though fairly simple) precedence criteria, which your implementation ignores. If we were actually following the Semantic Versioning specs, then there seem to be lots of libraries available for correctly parsing and comparing it, but it seems that we're stuck with the X.X.X.Y format because previous releases have used that scheme.

catong commented on VideoLocker/res/values/errors.xml:
```diff
@@ -63,6 +63,19 @@
+    <!-- Snackbar notification shown when the current version of the app is unsupported -->
+    <string name="app_version_unsupported">Your version of the app is no longer supported. Upgrade to the latest version.</string>
+    <!-- Snackbar notification shown when the current version of the app is deprecated -->
+    <string name="app_version_deprecated">Your version of the app is no longer supported. Upgrade to the latest version by {last_supported_date} to continue accessing course content online.</string>
```
Please change "Upgrade" to "Update" in the string.

catong commented on VideoLocker/res/values/errors.xml:
```diff
@@ -63,6 +63,19 @@
+    <!-- Snackbar notification shown when the current version of the app is deprecated -->
+    <string name="app_version_deprecated">Your version of the app is no longer supported. Upgrade to the latest version by {last_supported_date} to continue accessing course content online.</string>
+    <!-- Snackbar notification shown when the current version of the app is outdated -->
+    <string name="app_version_outdated">A new version of the app is available. Upgrade to the latest version.</string>
```
Please change "Upgrade" to "Update" in the string.

catong commented on VideoLocker/res/values/errors.xml:
```diff
@@ -63,6 +63,19 @@
 
+    <!-- New version available -->
+    <!-- Snackbar notification shown when the current version of the app is unsupported -->
+    <string name="app_version_unsupported">Your version of the app is no longer supported. Upgrade to the latest version.</string>
```
Please change "Upgrade" to "Update" in the string.

1zaman replied:
This change has already been done in commit eb6b4d8eb6ea93c9f5a4ea6d62f2f85cad589abb. You may have been looking at an older commit.

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/util/AppUpdateUtils.java:
```diff
@@ -9,49 +9,52 @@
     /**
-     * Open an app store to display the app, or a web browser displaying it's web page if no app
-     * store is available.
+     * Open an native app or website on a web browser to update the app.
```
s/an/a

1zaman replied:
You're commenting on an older version of this javadoc, but this error was still retained in the latest version. Now fixed in commit e53fdcf9a32adaf821326a1a217570b9bbb79df7.

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/util/Config.java:
```diff
@@ -60,6 +67,40 @@
+            return formatApplicationId(webUri);
+        }
+
+        private static String formatApplicationId(@NonNull final String uri) {
```
A javadoc here would be helpful explaining the format of the Uri and what this function does.

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/util/Config.java:
```diff
@@ -385,6 +389,25 @@ public String getOAuthClientId() {
     }
 
+    /**
+     * @return A list of URIs for updating the app, or an empty list if none are available.
```
Suggestion: A javadoc here would be helpful explaining the format of the Uri string/s.

1zaman replied:
The token replacement is being done internally in the method, so a public javadoc wouldn't be an appropriate venue to document it, and the code is pretty self-evident. Nonetheless, I have added an inline comment explaining it in commit 27ae618bd9896badef5bd4dd600d6431b076ede7.

miankhalid commented:
Great job done by standardizing the Http statuses in `HttpStatus.java` class.

miankhalid commented on VideoLocker/src/test/java/org/edx/mobile/util/VersionTest.java:
```diff
@@ -0,0 +1,143 @@
+     * minor, and patch version properties.
+     */
+    @Test
+    public void newInstance_withTwoDotRelease_isParsedCorrectly() {
```
Tests in this class are not following [Android's Guidelines for naming unit tests](https://source.android.com/source/code-style.html#javatests-style-rules).

bguertin replied:
Actually, I would say that it does; Since there is only 1 single guideline on that page: `Follow test method naming conventions and use an underscore to separate what is being tested from the specific case being tested. This style makes it easier to see exactly what cases are being tested.`

This method is doing exactly that.

miankhalid replied:
> If you are referring to the test prefix

Yeah, we've been following this prefix format for all our tests since we saw this guideline.

p.s. +1 on the fact that its totally redundant.

1zaman replied:
`test` prefix added in commit 07b1284c92f05a1d40ab8295c703fa5d94d7c361.

miankhalid commented on VideoLocker/src/test/java/org/edx/mobile/util/VersionTest.java:
```diff
@@ -0,0 +1,143 @@
+     * characters in one of the first three tokens throws a
+     * {@link NumberFormatException}.
+     */
+    @Test(expected = NumberFormatException.class)
```
Interesting, didn't knew one could test if exceptions are being thrown correctly 😉

miankhalid commented:
Done with my pass. Other than my comments LGTM 👍

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,204 @@
+        try {
+            event = new NewVersionAvailableEvent(newVersion, lastSupportedDate, isUnsupported);
+        } catch (IllegalArgumentException e) {
+            return;
```
Looks like you are hiding an error here. Could we perhaps have  a comment on why, and consider logging it as well (with Crashlytics report)?

1zaman replied:
So this static utility method was supposed to accept all possible inputs, but only post an event when a valid one could be constructed (thus suppressing these exceptions). This is documented in the javadoc as well. However, the validation logic was being duplicated and fragmented across the caller and the constructor. This has now been resolved in commit f44b06f4701cf6519f244e120d06676dba503ba6, which also includes a comment here affirming that suppression is the proper and expected behaviour here.

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/http/HttpStatus.java:
```diff
@@ -0,0 +1,288 @@
+/**
+ * A comprehensive collection of constants for HTTP status codes.
+ */
+public class HttpStatus {
```
Curious: Why not an enum?

1zaman replied:
Do we need to validate the status codes as well? Otherwise an enum would only complicate the comparison, as the status codes we get from the client would need to be converted into the appropriate enum type, while having to handle a null value or an exception in the case where it's not represented in this set.

bguertin commented:
🆒 Looks good, except for my comment about the swallowed exception

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/http/NewVersionBroadcastInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+                     * schema, then discard the data and just return the
+                     * response.
+                     */
+                    logger.error(e);
```
We should log this error to crashlytics, which does not happen by default, just pass `true` as the 2nd parameter.

1zaman replied:
Done in commit 517ac01021476458cc65925cd50afc740dca7ed0.

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/event/NewVersionAvailableEvent.java:
```diff
@@ -0,0 +1,241 @@
+                try {
+                    currentVersion = new Version(BuildConfig.VERSION_NAME);
+                } catch (ParseException e) {
+                    logger.error(e);
```
I think this should be logged to Crashlytics as well, since it would indicate broken functionality.

1zaman replied:
Done in commit 517ac01021476458cc65925cd50afc740dca7ed0.

bguertin commented:
:+1:

bguertin commented:
What!? When I clicked the checkbox next to my name in the PR description, the description somehow got replaced with the description from a different pull request!! (https://github.com/edx/edx-app-android/pull/823)

mdinino commented:
Beautiful. 👍

1zaman commented:
It's OK; the summary was getting more and more out of date with recent changes, and was long overdue for a rehash anyway. I'll just update it with the commit messages after squashing.

1zaman commented:
Comprehensive unit tests have been added for the event and interceptor in commit 54bf4afadeac0be9fa04bc4ad0baf64d6f643ebc.

Unfortunately, since we currently don't have a Robolectric test runner that supports parameterized testing (and RoboGuice injection also requires a `Context`), the event bus posting can't be tested at the moment (without having to run it in the emulator), since the event bus uses the `Looper`, which is only present in the Android framework. Therefore, those tests are being ignored for now.

No UI tests for the Snackbar display implementation have been written, since there is no official Robolectric shadow `Snackbar` at the moment, and I don't have the time to write/verify a custom one. If we do add a custom one at some point, then this should be a simple test to write.

1zaman commented:
This pull request has dragged on for far too long, so I'm merging it now. However, feel free to take a look at (and comment on) the tests that I added.

miankhalid commented:
![finally](https://cloud.githubusercontent.com/assets/1710804/18301757/516e4c36-74ed-11e6-8dd9-f8dd2975d714.jpeg)
